statescript/executor.go

```diff
@@ -179,8 +210,43 @@ func execute(name string, timeout time.Duration) int {
 	return 0
 }
 
+// Catches a script that requests a retry in a loop. Is limited by the total window given to a script demanding a
+// retry.
+func executeScript(s os.FileInfo, dir string, l Launcher, timeout time.Duration, ignoreError bool) error {
+
+	iet := time.Now()
+	for {
+		switch ret := execute(filepath.Join(dir, s.Name()), timeout); ret {
+		case 0:
+			// success
+			return nil
+		case exitRetryLater:
+			if ignoreError {
+				log.Errorf("statescript: ignoring error executing '%s': %d", s.Name(), ret)
+				return nil
+			}
+			if time.Since(iet) <= l.getRetryTimeout() {
+				log.Errorf("statescript: %s requested a retry", s.Name())
+				time.Sleep(l.getRetryInterval())
+				continue
+			}
+			return errors.New("statescript: retry time-limit exceeded")
+		default:
+			// In case of error scripts all should be executed.
+			if ignoreError {
+				log.Errorf("statescript: ignoring error executing '%s': %d", s.Name(), ret)
+				return nil
+			}
+			return errors.Errorf("statescript: error executing '%s': %d",
+				s.Name(), ret)
+		}
+	}
+}
```

Comment:
@pasinskim like we discussed the functionality has now been moved out into a function to hinder future breakage in case of misuse. Thus now a return is used to break out of the loop when needed, instead of other spurious hacks we've tried out previously.